src/Stratis.FederatedPeg.Features.FederationGateway/SourceChain/MaturedBlocksProvider.cs

```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Features.BlockStore;
+using Stratis.FederatedPeg.Features.FederationGateway.Interfaces;
+
+namespace Stratis.FederatedPeg.Features.FederationGateway.SourceChain
+{
+    public class MaturedBlocksProvider : IMaturedBlocksProvider
+    {
+        private ConcurrentChain chain;
+        private IDepositExtractor depositExtractor;
+        private IBlockRepository blockRepository;
+
+        public MaturedBlocksProvider(ILoggerFactory loggerFactory,
+            ConcurrentChain chain, IDepositExtractor depositExtractor, IBlockRepository blockRepository)
+        {
+            this.chain = chain;
+            this.depositExtractor = depositExtractor;
+            this.blockRepository = blockRepository;
+        }
+
+        /// <inheritdoc />
+        public async Task<List<IMaturedBlockDeposits>> GetMaturedDepositsAsync(int blockHeight, int maxBlocks)
+        {
+            int currentHeight = blockHeight;
+            int matureHeight = (this.chain.Tip.Height - (int)this.depositExtractor.MinimumDepositConfirmations);
+
+            if (currentHeight > matureHeight)
+            {
+                throw new Exception($"Block height {currentHeight} submitted is not mature enough. Blocks less than a height of {matureHeight} can be processed.");
```

Comment:
Should it be "Blocks less than a height of {matureHeight} can" or "cannot be processed" ?

Also is there a more specific exception type to throw or will Exception be good enough.

Reply:
"can be processed". The exception type is discarded by the calling code which builds the response. Nevertheless I changed it to `InvalidOperationException`.

Reply:
Thanks